c10/core/ScalarType.h

```diff
@@ -592,6 +626,12 @@ static inline ScalarType promoteTypes(ScalarType a, ScalarType b) {
     return ScalarType::Undefined;
   }
 
+  if (isBarebonesUnsignedType(a) && a == b) {
+    return a;
```

Comment:
Note that the current type promotion is such that if a == b, return a is fine. I expect this to remain true so we might want to refactor that check at the top for minimal speed bump (since it is the common case) and simplifying some of these conditions.

Reply:
Fixed post Driss rebase.